### PR TITLE
Support deepseek v4 and kimi k3 on ssd

### PR DESCRIPTION
**Pull request by beyondHJM**
# [Feature] Add SSD-backed Expert Pack inference for DeepSeek-V4-Flash and Kimi-K3

## Summary

This PR adds an opt-in SSD-backed inference path for **DeepSeek-V4-Flash** and **Kimi-K3**. It allows these MoE models to run when their complete weights are larger than GPU VRAM and host RAM combined.

Routed expert weights are stored on SSD and loaded only when selected by the router. The implementation uses:

- an expert-major **Expert Pack** layout for efficient random access;
- direct I/O into aligned pinned host buffers;
- asynchronous host-to-device transfers; and
- a byte-budgeted LFU/LRU GPU cache for frequently used experts.

The runtime does not prune, merge, skip, replace, or re-quantize experts, and does not reduce Expert Top-K. The default `auto`, `safetensors`, and `gguf` loading paths remain unchanged; SSD execution is enabled only through the explicit `expert_pack` path.

## Supported Scope

`ExpertPackModelLoader` is selected only with `--load-format expert_pack`. Official checkpoints loaded through the standard `auto` or `safetensors` paths never enter this loader.

The currently validated DeepSeek input is the single-file MXFP4 GGUF used by `frob/deepseek-v4-flash-0731:latest`. Its routed expert matrices use `MXFP4`; the large non-routed matrices are mostly `Q8_0`, router input weights are `BF16`, and the inventory also contains `F32` and `I32` normalization or auxiliary tensors. The inspected GGUF contains 129 `MXFP4`, 512 `Q8_0`, 43 `BF16`, 641 `F32`, and 3 `I32` tensors. It contains no FP8 tensors.

The other validated input is the text-only 38-shard `KIMI-K3-MXP4-DERISKED-Q2_K-*.gguf`, whose routed experts use Q2_K for gate/up and Q3_K for down. The Kimi launcher extracts the official `text_config` into a flat `KimiLinearConfig`; the complete multimodal Kimi-K3 checkpoint is not supported by this loader. Other GGUF quantization variants have not been validated and are not currently claimed as supported.

## Usage

Install this branch from source before running the examples:

```bash
python3 -m pip install -e "python"
```

### DeepSeek-V4-Flash

The tested MXFP4 GGUF is available as [`frob/deepseek-v4-flash-0731:latest`](https://ollama.com/frob/deepseek-v4-flash-0731), which was imported from [`unsloth/DeepSeek-V4-Flash-0731-GGUF`](https://huggingface.co/unsloth/DeepSeek-V4-Flash-0731-GGUF). The benchmark used the immutable Ollama model blob with SHA-256 `947ac34c08c0e5c5752ac76398f934b3b6b4075cfe915ba43dd5ac754900a4cd`.

```bash
python3 examples/runtime/deepseek_v4/benchmark_deepseek_5090.py \
  --gguf /path/to/deepseek-v4-flash-0731-mxfp4.gguf \
  --prompt "Explain why the sky appears blue." \
  --max-new-tokens 200
```

### Kimi-K3

Download all 38 Q2_K GGUF shards from [`Blackfrost-AI/KIMI-K3-Q2_K-GGUF-ABLITERATED`](https://huggingface.co/Blackfrost-AI/KIMI-K3-Q2_K-GGUF-ABLITERATED). Also place the tokenizer/config files from [`moonshotai/Kimi-K3`](https://huggingface.co/moonshotai/Kimi-K3) in a sibling directory named `kimi-k3-tokenizer`.

```bash
hf download Blackfrost-AI/KIMI-K3-Q2_K-GGUF-ABLITERATED \
  --include "KIMI-K3-MXP4-DERISKED-Q2_K-*.gguf" \
  --local-dir /path/to/kimi-k3
```

```bash
python3 examples/runtime/kimi_k3/benchmark_kimi_k3_5090.py \
  --gguf /path/to/kimi-k3/KIMI-K3-MXP4-DERISKED-Q2_K-00001-of-00038.gguf \
  --prompt "Explain why the sky appears blue." \
  --max-new-tokens 200
```

The first shard is only the entrypoint. The launcher automatically discovers and uses all 38 shards in the same directory. Both launchers derive the remaining paths automatically. If the Expert Pack is missing, it is built on the first launch and reused afterward.

| Model | First Expert Pack build | Approximate first-run readiness |
| --- | ---: | ---: |
| DeepSeek-V4-Flash | Estimated **5-10 minutes** on the benchmark NVMe SSD | **8-15 minutes** |
| Kimi-K3 | **29 minutes 42 seconds measured** | **35-45 minutes** |

The 38 Kimi-K3 source shards and generated Expert Pack occupy approximately 1.814 TiB in total, so a 4 TB SSD is recommended for practical headroom.

## Correctness

- DeepSeek-V4-Flash produced semantically equivalent answers to Ollama across travel planning, science, history, programming, and mathematics prompts.
- Kimi-K3 matched the 200-token SGLang reference output exactly. All 92 routed layers executed Top-16 experts with `fallback_count=0` and `io_errors=0`.

## Performance

The target platform is one RTX 5090 with 32 GB VRAM, 32 GB host RAM, and NVMe SSD storage. The retained measurements were collected on RTX 5090 hosts; host RAM and SSD capacity differed between the two model runs.

### DeepSeek-V4-Flash vs Ollama

Five prompts were measured with 200 generated tokens per prompt.

| Metric | SGLang | Ollama | Result |
| --- | ---: | ---: | ---: |
| Prefill | 3.265 token/s | 1.457 token/s | **2.24x faster** |
| Decode | 2.020 token/s | 1.354 token/s | **1.49x faster** |

### Kimi-K3 vs llama.cpp

The SGLang run used a 91-token prompt and generated 200 tokens.

| Metric | SGLang | llama.cpp | Result |
| --- | ---: | ---: | ---: |
| Prefill | 0.66412 token/s | 0.11289 token/s | **5.88x faster** |
| Decode | 0.44830 token/s | 0.1810 token/s | **2.48x faster** |

The llama.cpp decode baseline came from an earlier 50-token run, so the 2.48x decode figure is indicative rather than a strict equal-length comparison.

---
### CI States

Latest PR Test (Base): :x: [Run #32837312025](https://github.com/sgl-project/sglang/actions/runs/32837312025)
Latest PR Test (Extra): :x: [Run #32837311604](https://github.com/sgl-project/sglang/actions/runs/32837311604)
Latest PR Test (AMD ROCm 7.2): :x: [Run #32837311774](https://github.com/sgl-project/sglang/actions/runs/32837311774)

### CODE REVIEW

File: python/sglang/srt/model_loader/expert_pack_loader.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+    def download_model(self, model_config) -> None:
+        if not Path(model_config.model_path).is_dir():
+            raise ValueError(
+                "expert_pack model_path must be the verified tokenizer/config directory"
+            )
+
+    def load_model(self, *, model_config, device_config) -> nn.Module:
+        hf_config = model_config.hf_config
+        is_kimi = hf_config.model_type == "kimi_linear"
```
Comment (BBuf) [inline python/sglang/srt/model_loader/expert_pack_loader.py]: `moonshotai/Kimi-K3` has `model_type: "kimi_k3"` at the top level - `kimi_linear` only appears in the nested `text_config`. So the official checkpoint falls through to the `!= "deepseek_v4"` check and gets "expert_pack supports only DeepSeek-V4 and Kimi-K3", which is a misleading error.

The hard constraints below have the same problem: `getattr(hf_config, "num_hidden_layers")` is None on the nested config, so it dies on `num_hidden_layers must be 93`.

Is the flat GGUF-converted config the only intended input? If so, please state that in the PR body. Reading `model_config.hf_text_config` would cover both layouts.

File: python/sglang/srt/layers/quantization/gguf.py [hunk 2]
```diff
@@ -71,6 +71,22 @@
...
+    if not shard_ids:
+        return []
+    if all(isinstance(shard_id, int) for shard_id in shard_ids):
+        expected = set(range(len(shard_ids)))
+        if len(set(shard_ids)) != len(shard_ids) or set(shard_ids) != expected:
+            raise ValueError(f"Invalid GGUF merged shard IDs: {shard_ids}")
+        return sorted(shard_ids)
+    if "q" in shard_ids:
+        if len(shard_ids) != 3 or set(shard_ids) != {"q", "k", "v"}:
+            raise ValueError(f"Invalid GGUF QKV shard IDs: {shard_ids}")
```
Comment (BBuf) [inline python/sglang/srt/layers/quantization/gguf.py]: This turns a previously-working load into a hard failure on the shared GGUF path. The old code was `shard_id = ["q","k","v"] if "q" in shard_id else shard_id`, which passed any other set through; now a fused layer carrying only `{q,k}` raises. Same at L81 for non-contiguous integer shard ids.

The offset change itself looks correct - the old `start` was accumulated in container order while `shard_offset_map` was keyed by idx, so offsets were wrong whenever the two orders differed. But it changes behavior for every GGUF model, not just expert_pack.

Did you load a non-DeepSeek/K3 GGUF model to confirm no behavior change? Please split this out with that regression, same question that came up in #17883.

File: python/sglang/srt/layers/attn_residual.py [hunk 3]
```diff
@@ -33,15 +33,19 @@
 _HIP_SHAPE_GATE = None
 
 
+def _supports_attn_res_tma(capability: tuple[int, int]) -> bool:
+    """Return whether the device implements the SM10x tcgen05/TMEM ISA."""
+    major, _ = capability
+    return major == 10
```
Comment (BBuf) [inline python/sglang/srt/layers/attn_residual.py]: `major >= 10` -> `major == 10` is a real fix for SM120, but it also silently drops the fast path on SM11x and anything later. No log, so a user cannot tell which backend they ended up on. Please log the selection once.

This is a hot kernel and the change has nothing to do with SSD offload - can you pull it into its own PR? A standalone "attn_res TMA gate excludes SM120" would land quickly. `test_attn_residual.py` only checks the pure-function mapping; it does not assert that SM10x still takes the TMA path.

File: python/sglang/srt/server_args.py [hunk 4]
```diff
@@ -123,6 +123,7 @@
     "sharded_state",
     "presharded",
     "gguf",
+    "expert_pack",
```
Comment (BBuf) [inline python/sglang/srt/server_args.py:133]: `--load-format expert_pack` on its own can never work. It also needs `model_loader_extra_config.pack_path` (or `SGLANG_EXPERT_PACK_PATH`), `manifest_path` for Kimi, four sha256 fields for DeepSeek, plus `--disable-cuda-graph`, `--disable-shared-experts-fusion` and TP=EP=1. None of the examples use this flag.

Same shape as the `ipc_cache` discussion in #27139. Either drop it from `LOAD_FORMAT_CHOICES` and keep the enum for internal selection, or validate in `ServerArgs.__post_init__` so the user gets one complete error at startup instead of discovering the requirements one raise at a time.

**Review by BBuf — COMMENTED** [review]
Nice work - the explicit hard constraints and the `--disable-cuda-graph` requirement are the right call for this kind of offload, and returning parameter coverage from `load_weights` to assert nothing was silently skipped is good.

Two things before this can get a proper review:

- CI is red on both lanes.
- There is no accuracy number for either model. The K3 numbers come from a third-party Q2_K abliterated GGUF, and "matched the SGLang reference output exactly" is a self-comparison against the same weights, so it says nothing about accuracy. A small GSM8K run through the expert_pack path would settle it. The PR also ships `expert_pack_mxfp4.cu` but no mxfp4 measurement - which test covers that kernel?

I would also split the two shared-path fixes (`attn_residual.py` SM gate, `gguf.py` shard ordering) into their own PRs. Both are independently useful and independently reviewable; right now they are buried in +8646 and can't be benchmarked separately.

File: python/sglang/srt/model_loader/expert_pack_loader.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+                expected_top_k=16,
+                cache_vram_mib=int(self.config.get("cache_vram_mib", 4 * 1024)),
+                cache_vram_reserve_mib=int(
+                    self.config.get("cache_vram_reserve_mib", 2 * 1024)
+                ),
+                stage_slots=int(self.config.get("stage_slots", 16)),
+                read_splits=int(self.config.get("read_splits", 4)),
+                direct_io=bool(self.config.get("direct_io", False)),
+                stats_flush_interval=int(self.config.get("stats_flush_interval", 0)),
+                verify_pack_sha256=bool(self.config.get("verify_pack_sha256", False)),
```
Comment (BBuf) [inline python/sglang/srt/model_loader/expert_pack_loader.py]: Kimi defaults `verify_pack_sha256` to False here, DeepSeek defaults the same key to True at L280. Intentional?

The pack is the artifact that gets randomly read at inference time, so corruption there surfaces as wrong experts with no error - exactly what `fallback_count` / `io_errors` exist to catch. Verifying the header and manifest digest on open would be cheap even if the full sha256 stays opt-in.

File: python/sglang/srt/layers/attn_residual.py (hunk 3 above)
Comment (beyondHJM) [reply python/sglang/srt/layers/attn_residual.py]: Thank you for the suggestion. I have updated the capability gate to preserve the fast path on SM11x and future architectures while excluding SM12x:
major >= 10 and major != 12
I have moved this change into a separate PR so that it can be reviewed and landed independently from the SSD offload changes:https://github.com/sgl-project/sglang/pull/35361. This PR modifies only four lines of code. Its CI checks have not passed yet; both the Base and Extra CI runs are currently failing, I don't know why.

File: python/sglang/srt/model_loader/expert_pack_loader.py (hunk 1 above)
Comment (beyondHJM) [reply python/sglang/srt/model_loader/expert_pack_loader.py]: A few clarifications on our use case:

1. We are using a quantized version of Kimi K3that only contains the language model part. It does **not** include the vision model. Because the vision component is removed, there is no longer a need for the nested `text_config` structure. The configuration is therefore flattened to the top level. In this quantized checkpoint, the `hf_config` is semantically identical to the `hf_text_config` of the official `moonshotai/Kimi-K3` checkpoint. So the current logic (`model_type == "kimi_k3"`) works correctly for our intended input.

2. More importantly, this PR's feature (the `--load-format expert-pack` path) is not triggered at all for the official `moonshotai/Kimi-K3` checkpoint. The official Kimi K3 release uses the standard safetensors format, so users would load it via the default `--load-format auto` (or `safetensors`) path, not `expert-pack`. The expert-pack loader is specifically designed for quantized checkpoints that package expert weights in a custom pack format. Therefore, the nested config issue you raised will not actually affect the feature's intended execution path.

File: python/sglang/srt/layers/quantization/gguf.py (hunk 2 above)
Comment (beyondHJM) [reply python/sglang/srt/layers/quantization/gguf.py]: Thanks, this is a great catch. I agree that partial QK layouts and non‑contiguous integer shard IDs must retain the previous pass‑through behavior.

I have reworked the logic in `gguf.py` and validated it across multiple models, including Qwen and Llama families. No loading errors were observed, and more importantly, there are no output differences and no performance differences compared to the original implementation. This demonstrates that the current `gguf.py` works correctly for both the `expert‑pack` path and the standard GGUF path.

I don't think this fix needs to be split into a separate PR, because it is directly tied to the `expert‑pack` feature, because the original `gguf.py` already failed to handle `expert‑pack` loading properly, as I confirmed during my testing.

Therefore, I prefer to keep the changes to `gguf.py` in this PR rather than reverting them.

File: python/sglang/srt/server_args.py (hunk 4 above)
Comment (beyondHJM) [reply python/sglang/srt/server_args.py:133]: Thanks, this was valid for the reviewed revision. We took the second approach and now handle  
`expert_pack` centrally during `ServerArgs` startup.

For a supported raw GGUF, the public invocation is now  
`--model-path <raw-gguf> --load-format expert_pack`. Server startup automatically prepares/resolves  
the tokenizer metadata, Expert Pack, manifest, and DeepSeek integrity fields. It also disables CUDA  
graph and shared-expert fusion automatically and validates TP/DP/EP and incompatible options in  
one aggregated startup error.

File: python/sglang/srt/model_loader/expert_pack_loader.py (hunk 5 above)
Comment (beyondHJM) [reply python/sglang/srt/model_loader/expert_pack_loader.py]: Thank you for catching this. Full payload hashing is no longer exposed through the serving configuration because scanning a nearly 1 TB pack on every startup is too expensive. It remains available as an explicit offline check.

Startup still performs mandatory structural validation. DeepSeek checks its manifest, pack size, header, identity digests, entries, and object layout. Kimi uses its own validator for shard inventory, tensor shapes and quantization, pack header/index, offsets, alignment, index SHA-256, and sampled payloads. The runtime stores validate the model-specific format again before inference.

The current branch passed 20-token smoke tests for both Kimi-K3 and DeepSeek-V4-Flash with no expert-pack fallback or I/O errors.

**Review by BBuf — APPROVED** [review]

**Comment by BBuf** [comment]
https://github.com/sgl-project/sglang/actions/runs/32837312025/job/97803698788?pr=35314

**Comment by merrymercy** [comment]
Style comment:

A top-level folder `tools/expert_pack` just for this pr is not allowed.
Can you move this folder to another place? It can be under scripts or your custom stuff or srt/utils

@beyondHJM  @BBuf can you follow up on this?

**Comment by cyxlily** [comment]
Heads up — this PR accidentally reverted the XPU MHC dispatch that #32166 added to python/sglang/srt/models/deepseek_v4.py‎ (likely a stale-branch merge). Filed #36803 to restore it. Please double check for similar silent reverts in future merges, thanks!

**Comment by 253844077** [comment]
Is there any deployment that implements hybrid inference for DS4.1 across memory, SSD, and GPU — putting Engram in memory, MoE on SSD, and activated experts on GPU? I feel this could be done with just 512GB of memory, a 2TB SSD, and 48GB of VRAM.
